src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheck.java

```diff
@@ -117,36 +185,167 @@ public void visitToken(DetailAST ast) {
         }
     }
 
+    /**
+     * Setter to custom overload groups, allows for the definitions of multiple overloading
+     * method groups.
+     *
+     * @param modifierGroupRegex the array of regex patterns
+     */
+    public void setModifierGroups(String modifierGroupRegex) {
+        final String[] split = modifierGroupRegex.split(",");
+        modifierGroups = new Pattern[split.length];
+        for (int index = 0; index < split.length; index++) {
+            final String trimmed = split[index].trim();
+            final Pattern pattern = Pattern.compile(trimmed);
+            modifierGroups[index] = pattern;
+        }
+    }
+
     /**
      * Checks that if overload methods are grouped together they should not be
      * separated from each other.
      *
-     * @param objectBlock
-     *        is a class, interface or enum object block.
+     * @param objectBlock is a class, interface or enum object block.
      */
     private void checkOverloadMethodsGrouping(DetailAST objectBlock) {
         final int allowedDistance = 1;
+        final Map<String, IndexAndLine> knownMethods = new HashMap<>();
```

Comment:
Can we use a list of `MethodDetails`as proposed below?

Reply:
Before I make changes, just to confirm, you are looking for something like:
```java
List<MethodDetails> knownMethods = new ArrayList<>();
for(MethodDetails md: knownMethods) {
    if(md.getMethodName().equals(thisMethodName) && md.getModifiers().equals(thisMethodsModifiers) {
          // confirm method is immediately after previous match
    } else {
         // haven't seen this collection of modifiers yet, add to our list
    }
}
```

Do you think going from O(1) lookup time to O(n) will be a concern?

Reply:
@austinarbor this is a good point; I like the map implementation, but I am not a big fan of the "description string". This string seems difficult to reason about/ debug in future maintenance. What do you think about `Map<MethodDetails, Integer> knownMethods`, where integer is the index and we store the line number in `MethodDetails`?

Reply:
@nrmancuso I had something similar to `Map<MethodDetails, Integer>` when I first opened the PR, but that creates other issues with code coverage as mentioned [here](https://github.com/checkstyle/checkstyle/pull/12086#discussion_r953206083). If we create a `private static` inner class and use it as a key in the map, we need to override `equals` and `hashcode`. 
1. PMD requires `null` check in the equals method of the class
2. It's impossible in the code for the value to ever be null, so we can't test this scenario via normal processing
3. Because the inner class is private we can't write unit test for this scenario without using reflection

Does that make sense?

We can get code coverage to pass if we make the `MethodDetails` a public class outside of the check, (because then we can write unit tests on the public method), but I didn't know where there would be a good place to put it. I also tried to make the inner class `protected` / `package protected` so I can call `equals` from the test class, but that caused other checkstyle checks to fail as well becauase that visibility was disallowed.

Reply:
Yes, it has been awhile since we looked at first iteration, thanks for the reminder. One last thought; what do you think of using a simple data class that contains the same information? We could do `Map<MethodDetails, IndexAndLine>` where `MethodDetails` is comprised of the pattern and method name?

Reply:
Hmm..wouldn't that have the same issue since `MethodDetails` is being used as a key?

Reply:
It shouldn’t, since we do not need to override equals/hashcode in this case, right?

Reply:
Why wouldn't we? We would need to override for any class that's used a key in the map otherwise it will use a referential equality check. In java 17 then we could just use a record, but there isn't a way around overriding equals and hashcode that I'm aware of for java 11....unless there's a trick I don't know about?

Reply:
>Why wouldn't we?

I was thinking of some hack with custom map, but this is not good either. Let's proceed with custom description string then, we can get all test cases and behavior of update correct first and give us some time to think on this.